test/typeEvaluate.test.ts

```diff
@@ -2092,6 +2093,84 @@ t.test('function: references', (t) => {
   t.end()
 })
 
+t.test('optimizations: hoistDuplicateUnionTypeNodes', (t) => {
+  const optimized = optimizeUnions({
+    type: 'union',
+    of: [
+      {
+        type: 'union',
+        of: [
+          {
+            type: 'string',
+            value: 'foo',
+          },
+          {
+            type: 'string',
+            value: 'bar',
+          },
+          {
+            type: 'number',
+            value: 3,
+          },
+        ],
+      },
+      {
+        type: 'union',
+        of: [
+          {
+            type: 'string',
+            value: 'foo',
+          },
+          {
+            type: 'string',
+            value: 'baz',
+          },
+          {
+            type: 'number',
+            value: 2,
+          },
+        ],
+      },
+    ],
+  })
+  t.strictSame(optimized, {
+    type: 'union',
+    of: [
+      {
+        type: 'string',
+        value: 'foo',
+      },
+      {
+        type: 'union',
+        of: [
+          {
+            type: 'number',
+            value: 2,
+          },
+          {
+            type: 'string',
+            value: 'baz',
+          },
+        ],
+      },
+      {
+        type: 'union',
+        of: [
+          {
+            type: 'number',
+            value: 3,
+          },
+          {
+            type: 'string',
+            value: 'bar',
+          },
+        ],
+      },
+    ],
+  })
```

Comment:
So, the original expression is `("foo" | "bar" | 3) | ("foo" | "baz" | 2 )`, and this asserts that optimizing it should become `"foo" | ("bar" | 3) | ("baz" | 2 )`. However, why don't we want this to become `("foo" | "bar" | 3 | "baz" | 2 )`? Isn't that also equivalent?